elements/bulbs-dfp/bulbs-dfp.js

```diff
@@ -0,0 +1,122 @@
+import util from 'bulbs-elements/util';
+import invariant from 'invariant';
+import {
+  registerElement,
+  BulbsHTMLElement,
+} from 'bulbs-elements/register';
+
+// We have to do this little dance to properly subclass elements in Safari
+function BulbsHTMLDivElement () {}
+BulbsHTMLDivElement.prototype = HTMLDivElement.prototype;
+
+export default class BulbsDfp extends BulbsHTMLElement {
+  attachedCallback () {
+    invariant(this.hasAttribute('refresh-interval'),
+        '<div is="bulbs-dfp>" MUST specify a `refresh-interval` attribute in ms');
+
+    invariant(this.hasAttribute('viewport-threshold'),
+        '<div is="bulbs-dfp"> MUST specify a `viewport-threshold` property. ' +
+        'This value defines how many screens away from the viewport the slot ' +
+        'will be in order to trigger an ad load. 1.2 = 120% viewport height.');
+
+    this.handleEnterViewport = this.handleEnterViewport.bind(this);
+    this.handleExitViewport = this.handleExitViewport.bind(this);
+    this.handleInterval = this.handleInterval.bind(this);
+
+    this.addEventListener('enterviewport', this.handleEnterViewport);
+    this.addEventListener('exitviewport', this.handleExitViewport);
+
+    let threshold = parseFloat(this.getAttribute('viewport-threshold'), 10);
+    util.InViewMonitor.add(this, {
+      get distanceFromTop () {
+        return window.innerHeight * threshold;
+      },
+      get distanceFromBottom () {
+        return -(window.innerHeight * threshold);
+      },
+    });
+
+    let intervalLength = parseFloat(this.getAttribute('refresh-interval', 10));
+    this.refreshInterval = window.setInterval(this.handleInterval, intervalLength);
+  }
+
+  detachedCallback () {
+    util.InViewMonitor.remove(this);
+    window.clearInterval(this.refreshInterval);
+  }
+
+  handleEnterViewport () {
+    util.getAnalyticsManager().sendEvent({
+      eventCategory: 'AdTech Metrics',
+      eventAction: 'enterviewport',
+      eventLabel: this.dataset.adUnit,
```

Comment:
Emitting an event for `enterviewport` so we'll get a sense of how many ads will eventually be kicked off by this element.